docs/source/cpp/tutorials/io_tutorial.rst

```diff
@@ -0,0 +1,404 @@
+data structures:
+
+.. literalinclude:: ../../../../cpp/examples/tutorial_examples/file_access_example.cc
+  :language: cpp
+  :start-after: (Doc section: Main)
+  :end-before: (Doc section: Main)
+
+Which, like when we used it before, is paired with a ``RunMain()``:
+
+.. literalinclude:: ../../../../cpp/examples/tutorial_examples/file_access_example.cc
+  :language: cpp
+  :start-after: (Doc section: RunMain)
+  :end-before: (Doc section: RunMain)
+
+Generating Files for Reading
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We need some files to actually play with. In practice, you’ll likely
+have some input for your own application. Here, however, we want to
+explore doing I/O for the sake of it, so let’s generate some files to make
+this easy to follow. To create those, we’ll define a helper function
+that we’ll run first. Feel free to read through this, but the concepts
+used will be explained later in this article. Note that we’re using the
+day/month/year data from the previous tutorial. For now, just copy the
+function in:
+
+.. literalinclude:: ../../../../cpp/examples/tutorial_examples/file_access_example.cc
+  :language: cpp
+  :start-after: (Doc section: GenInitialFile)
+  :end-before: (Doc section: GenInitialFile)
+
+To get the files for the rest of your code to function, make sure to
+call ``GenInitialFile()`` as the very first line in ``RunMain()`` to initialize
+the environment:
+
+.. literalinclude:: ../../../../cpp/examples/tutorial_examples/file_access_example.cc
+  :language: cpp
+  :start-after: (Doc section: Gen Files)
+  :end-before: (Doc section: Gen Files)
+
+I/O with Arrow Files
+--------------------
+
+We’re going to go through this step by step, reading then writing, as
+follows:
+
+1. Reading a file
+
+   a. Open the file
+
+   b. Bind file to :class:`ipc::RecordBatchFileReader`
+
+   c. Read file to :class:`RecordBatch`
+
+2. Writing a file
+
+   a. Get a :class:`io::FileOutputStream`
+
+   b. Write to file from :class:`RecordBatch`
+
+Opening a File
+^^^^^^^^^^^^^^
+
+To actually read a file, we need to get some sort of way to point to it.
+In Arrow, that means we’re going to get a :class:`io::ReadableFile` object – much
+like an :class:`ArrayBuilder` can clear and make new arrays, we can reassign this
+to new files, so we’ll use this instance throughout the examples:
+
+.. literalinclude:: ../../../../cpp/examples/tutorial_examples/file_access_example.cc
+  :language: cpp
+  :start-after: (Doc section: ReadableFile Definition)
+  :end-before: (Doc section: ReadableFile Definition)
+
+A :class:`io::ReadableFile` does little alone – we actually have it bind to a file
+with :func:`io::ReadableFile::Open`. For
+our purposes here, the default arguments suffice:
+
+.. literalinclude:: ../../../../cpp/examples/tutorial_examples/file_access_example.cc
+  :language: cpp
+  :start-after: (Doc section: Arrow ReadableFile Open)
+  :end-before: (Doc section: Arrow ReadableFile Open)
+
+Opening an Arrow file Reader
```

Comment:
Based on this ML discussion we may have to circle back here and make the terminology consistent, but this is OK for now https://lists.apache.org/thread/d2l7179yd41gfdhktrl1h06gh31l8l4l

Reply:
I'm going with Arrow files for now, and will swap in Arrow IPC if it becomes solidified.